lib/cerner/oauth1a/access_token.rb

```diff
@@ -174,14 +182,21 @@ def authorization_header
       def authenticate(access_token_agent)
         raise ArgumentError, 'access_token_agent is nil' unless access_token_agent
 
+        if @realm && !@realm.eql?(access_token_agent.realm)
+          raise OAuthError.new('realm does not match provider', nil, 'realm_rejected', nil, access_token_agent.realm)
```

Comment:
'realm_rejected' is not actually allowed as an 'oauth_problem' according to http://wiki.oauth.net/w/page/12238543/ProblemReporting: "Other values of oauth_problem MUST NOT be used."

Reply:
I'm thinking the 'oauth_problem' should just be nil, and the WWW-Authenticate header sent back to consumers would contain the correct realm. It will be up to consumers to identify the mismatch.

Providers may also include the OAuthError message in the body for additional context.

Reply:
The 'realm' parameter is outside of the token since it's actually from [rfc7235](https://tools.ietf.org/html/rfc7235), but this might actually be a good option. It may very well be a valid access token, but it's not for the realm supported by the resource server, so it's being rejected. That seems pretty appropriate for this case.

I've read through a lot of oauth1 libraries on github and most of them are pretty bad. More than half of them don't have any handling of the 'realm' parameter at all. Many don't send/build WWW-Authenticate headers at all. I haven't found any that both validate the access token realm against the provider realm and build a WWW-Authenticate header.

The python [oauthlib](https://github.com/oauthlib/oauthlib/blob/371029906aa6ccc9943120e096a7bbdd0ef6945d/docs/oauth1/server.rst) does realm validation, but doesn't help with building the proper WWW-Authenticate header.

Reply:
Converted to 'token_rejected': https://github.com/cerner/cerner-oauth1a/pull/4/commits/c5f60b0237192f5f97cf56726968169df4c97867